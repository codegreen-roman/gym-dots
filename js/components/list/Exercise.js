import React from 'react'
import glamorous from 'glamorous'
import { shape, string, number } from 'prop-types'

const _Exercise = ({ exercise }) => {
    const ExerciseTitle = glamorous.div({
        fontSize: 20,
        fontWeight: 'bold',
        color: '#606b7b'
    })
    const ExerciseDetails = glamorous.div({
        fontSize: 15,
        color: '#929aa6'
    })

    return (
        <div>
            <ExerciseTitle>{exercise.name}</ExerciseTitle>
            <ExerciseDetails>{exercise.weight} kg x {exercise.reps} reps x {exercise.setsLeft} sets</ExerciseDetails>
        </div>
    )
}

_Exercise.propTypes = {
    exercise: shape({
        name: string.isRequired,
        weight: number.isRequired,
        reps: number.isRequired,
        sets: number.isRequired,
    })
}

export const Exercise = _Exercise
