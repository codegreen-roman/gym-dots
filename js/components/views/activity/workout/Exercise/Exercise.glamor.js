import { css } from 'glamor'

export const restingTimer = isMoreThanRestTime =>
    css({
        fontSize: '110%',
        color: isMoreThanRestTime ? '#DC143C' : '#006400'
    })
