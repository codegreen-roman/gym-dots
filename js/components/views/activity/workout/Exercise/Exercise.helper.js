export function fakeExercise() {
    return {
        name: 'Monday Sweat',
        status: 'started',
        restTime: 60,
        sets: 5,
        setsLeft: 5,
        reps: 12,
        weight: 100,
        results: []
    }
}
