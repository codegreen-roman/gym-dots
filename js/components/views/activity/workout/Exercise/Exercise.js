import { Observable, Subject } from 'rxjs'
import React from 'react'
import { equals, compose, not, inc } from 'ramda'
import { string, arrayOf, number, bool } from 'prop-types'
import { CircleProgress } from '../../../../common/circle/CircleProgress'

const compare = compose(not, equals)
const isResultDiff = (oldResults, newResults) => {
    return compare(
        JSON.stringify(oldResults),
        JSON.stringify(newResults)
    )
}

export class Exercise extends React.Component {

    static propTypes = {
        status: string.isRequired,
        restTime: number.isRequired,
        sets: number.isRequired,
        setsLeft: number.isRequired,
        reps: number.isRequired,
        weight: number.isRequired,
        results: arrayOf(bool).isRequired
    }

    state = {
        seconds: 0
    }

    constructor(props) {
        super(props)

        const sideEffect = seconds => {
            this.setState({
                seconds
            })
        }

        this.restingStart$ = new Subject()
        this.kill$ = new Subject()

        this.restingStart$
            .switchMap(() => Observable.timer(0, 1000)
                .take(this.props.restTime + 20))
            .map(inc)
            .takeUntil(this.kill$)
            .subscribe(sideEffect)

    }

    componentWillUnmount() {
        this.kill$.next()
    }

    componentWillUpdate({results : newResults}) {

        const { results : oldResults } = this.props

        if (isResultDiff(oldResults, newResults)) {
            this.restingStart$.next()
        }
    }

    renderRestingTimer = () => {

        const { seconds } = this.state
        const { restTime } = this.props
        const style = {
            fontSize: '110%',
            color: seconds > restTime ? '#DC143C' : '#006400'
        }

        if (this.state.seconds > 0) {
            return (<div style={style}>
                Resting for {this.state.seconds} / {restTime}
            </div>)
        }

        return null
    }

    render() {

        const {
            status,
            restTime,
            sets,
            setsLeft,
            reps,
            weight,
            results
        } = this.props

        return (
            <section>
                <div>
                    {status}
                </div>
                <div>
                    weight: {weight}
                </div>
                <div>
                    reps: {reps}
                </div>
                <div>
                    sets {sets - setsLeft} / {sets} done
                </div>
                <div>the time</div>
                <div>
                    results : {JSON.stringify(results, null, 2)}
                </div>
                <div>
                    setsLeft : {setsLeft}
                </div>
                <div>
                    <button>Make pause</button>
                </div>
                {this.renderRestingTimer()}
                <div>
                    {restTime} sec rest time
                </div>
                <CircleProgress sets={sets} results={results} setsDone={sets - setsLeft} />
            </section>
        )
    }
}
