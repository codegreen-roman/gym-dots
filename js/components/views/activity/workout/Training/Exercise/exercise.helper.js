import { date } from 'faker'

export function fakeExercise() {
    return {
        name: date.weekday(),
        status: 'started',
        restTime: 60,
        sets: 5,
        setsLeft: 5,
        reps: 12,
        weight: 100,
        results: []
    }
}
