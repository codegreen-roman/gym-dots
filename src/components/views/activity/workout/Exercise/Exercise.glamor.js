import { css } from 'glamor'

export const restingTimer = isMoreThanRestTime =>
    css({
        fontSize: '110%',
        color: isMoreThanRestTime ? '#DC143C' : '#006400'
    })


export const exerciseSection = css({
    position: 'relative'
})

export const exerciseHeader = css({
    display: 'flex',
    flexFlow: 'row wrap',
    justifyContent: 'center',
    padding: '20px 0'
})

export const exerciseBody = css({
    position: 'relative'
})


export const circleProgressWrapper = css({

})

export const exerciseFooter = css({

})

