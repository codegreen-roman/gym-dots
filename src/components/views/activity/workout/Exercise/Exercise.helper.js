import { equals, compose, not } from 'ramda'

export const fakeExercise = () => ({
    name: 'Monday Sweat',
    status: 'started',
    restTime: 60,
    sets: 5,
    setsLeft: 5,
    reps: 12,
    weight: 100,
    results: []
})

export const compare = compose(not, equals)
export const isResultDiff = (oldResults, newResults) => {
    return compare(
        JSON.stringify(oldResults),
        JSON.stringify(newResults)
    )
}
