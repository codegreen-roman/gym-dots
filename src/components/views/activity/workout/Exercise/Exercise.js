import { Observable, Subject } from 'rxjs'
import React from 'react'
import { equals, compose, not, inc } from 'ramda'
import { arrayOf, number, bool } from 'prop-types'
import { CircleProgress } from '../../../../common/circle/CircleProgress'
import { Widget } from '@components/common/widget/Widget'
import {
    restingTimer,
    exerciseSection,
    exerciseHeader,
    exerciseBody,
    exerciseFooter
} from './Exercise.glamor'

const compare = compose(not, equals)
const isResultDiff = (oldResults, newResults) => {
    return compare(
        JSON.stringify(oldResults),
        JSON.stringify(newResults)
    )
}

export class Exercise extends React.Component {

    static propTypes = {
        restTime: number.isRequired,
        sets: number.isRequired,
        setsLeft: number.isRequired,
        reps: number.isRequired,
        weight: number.isRequired,
        results: arrayOf(bool).isRequired
    }

    state = {
        seconds: 0
    }

    constructor(props) {
        super(props)

        const sideEffect = seconds => {
            this.setState({
                seconds
            })
        }

        this.restingStart$ = new Subject()
        this.kill$ = new Subject()

        this.restingStart$
            .switchMap(() => Observable.timer(0, 1000)
                .take(this.props.restTime + 20))
            .map(inc)
            .takeUntil(this.kill$)
            .subscribe(sideEffect)

    }

    componentWillUnmount() {
        this.kill$.next()
    }

    componentWillUpdate({results : newResults}) {

        const { results : oldResults } = this.props

        if (isResultDiff(oldResults, newResults)) {
            this.restingStart$.next()
        }
    }

    renderRestingTimer = () => {

        const { seconds } = this.state
        const { restTime } = this.props

        if (this.state.seconds > 0) {
            return (
                <div {...restingTimer(seconds > restTime)}>
                    Resting for {this.state.seconds} / {restTime}
                </div>)
        }

        return null
    }

    render() {

        const {
            sets,
            setsLeft,
            reps,
            weight,
            results
        } = this.props
        const setsDone = (sets - setsLeft) / sets
        return (
            <section {...exerciseSection}>
                <div {...exerciseHeader}>
                    <Widget
                        iconName='dumbbells'
                        iconColor='red'
                        dataNumber={weight}
                        dataUnits='kg'
                        viewBox='0 0 34 20'
                        style={{
                            width: 34,
                            height: 20
                        }}
                    />
                    <Widget
                        iconName='check'
                        iconColor='red'
                        dataNumber={reps}
                        dataUnits='reps'
                        viewBox='0 0 27 27'
                        style={{
                            width: 23,
                            height: 23
                        }}
                    />
                </div>
                <div {...exerciseBody}>
                    <CircleProgress
                        sets={sets}
                        results={results}
                        setsDone={sets - setsLeft}
                    >
                        <div>sets {setsDone} done </div>
                        {this.renderRestingTimer()}
                    </CircleProgress>

                </div>
                <div {...exerciseFooter}>
                    results : {JSON.stringify(results, null, 2)}
                </div>
            </section>
        )
    }
}
