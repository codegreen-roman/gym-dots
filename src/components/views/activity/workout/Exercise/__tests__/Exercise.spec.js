import React from 'react'
import { Exercise } from '../Exercise'
import { render } from 'enzyme'

const defaultProps = {
    status: 'started',
    exerciseId: '1111',
    restTime: 60,
    setsLeft: 5,
    sets: 5,
    reps: 8,
    weight: 100,
    results: []
}

const setup = (props = defaultProps) => {
    return {
        props,
        component: render(<Exercise {...props} />)
    }
}

describe('Exercise component', () => {
    it('should match the snapshot', () => {
        const { component } = setup()
        expect(component).toMatchSnapshot()
    })
})
