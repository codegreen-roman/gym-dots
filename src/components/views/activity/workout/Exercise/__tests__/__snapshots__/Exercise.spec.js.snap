// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exercise component should match the snapshot 1`] = `
.css-1xng80g,
[data-css-1xng80g] {
  position: relative;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-justify-content: center;
  -webkit-align-items: center;
}

.css-1qcxu9b,
[data-css-1qcxu9b] {
  position: absolute;
}

.css-4t1vl3,
[data-css-4t1vl3] {
  display: inline-block;
  vertical-align: middle;
}

.css-1c4gejy,
[data-css-1c4gejy] {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 0.8;
  padding: 0 15px;
  -webkit-flex-flow: row wrap;
  -webkit-justify-content: center;
}

.css-194amhy,
[data-css-194amhy] {
  margin-left: 10px;
}

.css-6xnw37,
[data-css-6xnw37] {
  font-size: 29px;
  display: inline-block;
  color: #919AA7;
}

.css-1co8uoe,
[data-css-1co8uoe] {
  font-size: 14px;
  display: inline-block;
  margin-left: 5px;
  color: #919AA7;
}

.css-13azwyo,
[data-css-13azwyo] {
  position: relative;
}

.css-1p3dylf,
[data-css-1p3dylf] {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 20px 0;
  -webkit-flex-flow: row wrap;
  -webkit-justify-content: center;
}

.css-1fpka8r,
[data-css-1fpka8r] {
  fill: red;
}

.css-8plr4h,
[data-css-8plr4h] {
  fill: none;
  stroke-width: 12px;
  stroke: lightgrey;
}

<section
  data-css-13azwyo=""
>
  <div
    data-css-1p3dylf=""
  >
    <div
      data-css-1c4gejy=""
    >
      <svg
        data-css-4t1vl3=""
        height="20"
        viewbox="0 0 34 20"
        width="34"
      >
        <path
          d="M33.48 6.796c-.353-.444-.85-.698-1.364-.698h-.773v-.653c0-1.173-.973-2.128-2.169-2.128h-1.018V2.112a2.15 2.15 0 0 0-.633-1.519c-.8-.79-2.183-.79-2.983 0a2.151 2.151 0 0 0-.634 1.52v5.654H10.094V2.113A2.15 2.15 0 0 0 9.46.594c-.8-.791-2.183-.791-2.983 0a2.149 2.149 0 0 0-.633 1.519v1.205H4.7a2.06 2.06 0 0 0-1.43.571c-.39.376-.615.9-.615 1.437v.772H1.89c-.516 0-1.016.255-1.37.7A2.516 2.516 0 0 0 0 8.362v3.257c0 .585.185 1.14.522 1.563.354.446.853.701 1.37.701h.764v.772c0 .537.224 1.06.616 1.437.385.368.892.571 1.43.571h1.142v1.205c0 .568.23 1.121.633 1.519.4.395.93.613 1.492.613a2.11 2.11 0 0 0 1.491-.613 2.15 2.15 0 0 0 .634-1.519v-5.653h13.812v5.653c0 .568.231 1.121.633 1.519.4.395.93.613 1.492.613a2.11 2.11 0 0 0 1.492-.613 2.15 2.15 0 0 0 .633-1.519v-1.205h1.018c1.196 0 2.17-.954 2.17-2.127v-.653h.772c.513 0 1.01-.255 1.364-.698a2.5 2.5 0 0 0 .52-1.557V8.353c0-.583-.185-1.136-.52-1.557zM1.89 12.77c-.202 0-.399-.107-.553-.3a1.384 1.384 0 0 1-.276-.852V8.362c0-.321.1-.632.276-.852.154-.193.35-.3.553-.3h.765v5.56H1.89zm2.81 2.78c-.267 0-.527-.102-.71-.278a.854.854 0 0 1-.272-.618V5.326c0-.23.096-.45.27-.617.186-.178.445-.28.712-.28h1.143v11.122H4.7zm4.33 2.317a.993.993 0 0 1-.3.71c-.398.394-1.126.394-1.524 0a.992.992 0 0 1-.3-.71V2.113c0-.267.106-.52.3-.71a1.091 1.091 0 0 1 1.524 0c.193.19.3.443.3.71v15.756zm1.063-6.765V8.878h13.812v2.225H10.094zm17 6.765a.993.993 0 0 1-.3.71c-.4.394-1.127.394-1.525 0a.992.992 0 0 1-.3-.71V2.112a.99.99 0 0 1 .3-.709 1.091 1.091 0 0 1 1.524 0c.194.19.3.443.3.71v15.755zm3.187-3.332c0 .56-.496 1.015-1.107 1.015h-1.018V4.43h1.018c.61 0 1.107.456 1.107 1.016v9.091zm2.656-2.909c0 .32-.1.627-.273.846-.153.192-.348.298-.548.298h-.773V7.21h.773c.2 0 .395.106.548.298.174.218.273.526.273.845v3.274z"
          data-css-1fpka8r=""
        />
      </svg>
      <div
        data-css-194amhy=""
      >
        <div
          data-css-6xnw37=""
        >
          100
        </div>
        <div
          data-css-1co8uoe=""
        >
          kg
        </div>
      </div>
    </div>
    <div
      data-css-1c4gejy=""
    >
      <svg
        data-css-4t1vl3=""
        height="23"
        viewbox="0 0 27 27"
        width="23"
      >
        <path
          d="M13.5 0c1.863 0 3.617.352 5.26 1.055a13.547 13.547 0 0 1 4.298 2.887 13.547 13.547 0 0 1 2.887 4.298C26.648 9.883 27 11.637 27 13.5s-.352 3.617-1.055 5.26a13.547 13.547 0 0 1-2.887 4.298 13.547 13.547 0 0 1-4.298 2.887C17.117 26.648 15.363 27 13.5 27s-3.617-.352-5.26-1.055a13.547 13.547 0 0 1-4.298-2.887 13.547 13.547 0 0 1-2.887-4.298C.352 17.117 0 15.363 0 13.5s.352-3.617 1.055-5.26a13.547 13.547 0 0 1 2.887-4.298A13.547 13.547 0 0 1 8.24 1.055C9.883.352 11.637 0 13.5 0zm0 25.339c1.635 0 3.168-.312 4.601-.936a11.921 11.921 0 0 0 3.744-2.545 12.197 12.197 0 0 0 2.531-3.757c.624-1.433.936-2.966.936-4.601 0-1.635-.312-3.168-.936-4.601a12.034 12.034 0 0 0-2.53-3.744A12.034 12.034 0 0 0 18.1 2.624C16.668 2 15.135 1.687 13.5 1.687c-1.635 0-3.168.313-4.601.937a12.034 12.034 0 0 0-3.744 2.53A12.034 12.034 0 0 0 2.624 8.9C2 10.332 1.687 11.865 1.687 13.5c0 1.635.313 3.168.937 4.601a12.197 12.197 0 0 0 2.53 3.757A11.921 11.921 0 0 0 8.9 24.403c1.433.624 2.966.936 4.601.936zm5.379-16.77a.834.834 0 0 1 .606-.263c.229 0 .431.088.607.263a.805.805 0 0 1 .237.594c0 .237-.079.435-.237.593L11.68 18.22l-.106.105a.765.765 0 0 1-.593.264.765.765 0 0 1-.593-.264L6.196 14.16a.805.805 0 0 1-.237-.593c0-.237.08-.435.237-.593a.834.834 0 0 1 .607-.264c.228 0 .43.088.606.264l3.56 3.56 7.91-7.964z"
          data-css-1fpka8r=""
        />
      </svg>
      <div
        data-css-194amhy=""
      >
        <div
          data-css-6xnw37=""
        >
          8
        </div>
        <div
          data-css-1co8uoe=""
        >
          reps
        </div>
      </div>
    </div>
  </div>
  <div
    data-css-13azwyo=""
  >
    <div
      data-css-1xng80g=""
    >
      <svg
        height="220"
        width="220"
      >
        <g>
          <path
            class="css-8plr4h"
            d="M 205.1056516 79.0983006 A 100 100 0 0 0 127.3648178 11.5192247"
          />
          <path
            class="css-8plr4h"
            d="M 168.7785252 190.9016994 A 100 100 0 0 0 209.0268069 96.0826899"
          />
          <path
            class="css-8plr4h"
            d="M 51.2214748 190.9016994 A 100 100 0 0 0 153.8371147 199.8794046"
          />
          <path
            class="css-8plr4h"
            d="M 14.8943484 79.0983006 A 100 100 0 0 0 38.06602 179.465837"
          />
          <path
            class="css-8plr4h"
            d="M 110 10 A 100 100 0 0 0 21.7052407 63.0528437"
          />
        </g>
      </svg>
      <div
        data-css-1qcxu9b=""
      >
        <div>
          sets 0 done 
        </div>
      </div>
    </div>
  </div>
  <div>
    results : []
  </div>
</section>
`;
